src/DispatchListener.php

```diff
@@ -61,11 +63,30 @@ public function attach(EventManagerInterface $events)
      */
     public function onDispatch(MvcEvent $e)
     {
-        $routeMatch       = $e->getRouteMatch();
+        $routeMatch     = $e->getRouteMatch();
+        $request        = $e->getRequest();
+        $application    = $e->getApplication();
+        $response       = $application->getResponse();
+        $serviceManager = $application->getServiceManager();
+
+        // middleware?
+        $middleware = $routeMatch->getParam('middleware', false);
+        if (false !== $middleware) {
+            if (is_string($middleware) && $serviceManager->has($middleware)) {
+                $middleware = $serviceManager->get($middleware);
+            }
+            if (!is_callable($middleware)) {
+                $middleware = is_string($middleware) ? $middleware : get_class($middleware);
+                $return = $this->marshalControllerNotFoundEvent($application::ERROR_MIDDLEWARE_CANNOT_DISPATCH, $middleware, $e, $application);
+                return $this->complete($return, $e);
+            }
+            $return = $middleware(Psr7Request::fromZend($request), Psr7Response::fromZend($response));
+            return $this->complete(Psr7Response::toZend($return), $e);
+        }
+
```

Comment:
I'd write the end logic as this:

```php
if (! $return instanceof \Psr\Http\Message\ResponseInterface) {
    $e->setResult($return);
    return $return;
}
$response = Psr7Response::toZend($return);
$e->setResult($response);
return $response;
```

which might fix the testing issue you're seeing.